### ISSUE TITLE
Host/Container management UI unable to fetch CPU/Memory/Network/Storage stats over ws

### ISSUE THREAD

**Issue by vrparmar** [rancher/rancher#3384]
Host/Container management UI unable to fetch CPU/Memory/Network/Storage stats over ws. 

On investing it seems that ws:// connection to the server fails with error 401 even if security is disable. 

![image](https://cloud.githubusercontent.com/assets/4142248/12517749/52358e4e-c0e9-11e5-9c9e-61fcaac32196.png)

**Comment by deniseschannon** [comment]
Can you provide some more details on your setup?

Do you have single Rancher server or multinodes (HA)? What version of Rancher are you running? 

It looks like this is a local setup (based on the OS being Boot2Docker. I'm not sure if that would factor into it and would need to try it out myself. I could probably look into it later this week or next if you provide me with more details so I can try to reproduce.

**Comment by vrparmar** [comment]
I'm running Rancher server Ranchee-v0.51.0, Cattle-v0.130.0, User Interface-v0.78.0 & Rancher Compose-v0.7.0 as a container on docker-machine as single node with disable access control.
Agent is running on a docker host running on bare metal.

**Comment by sfarges** [comment]
Hello, got same issue websocket 401.

Host1: fresh install of Ubuntu 14.04.1 LTS (with KVM) (3.13.0-44-generic)
sudo docker run -d --restart=always --name Rancher_Server -p 8080:8080 rancher/server

Host2: fresh install of Ubuntu 14.04.1 LTS (with KVM) (3.13.0-44-generic)
install docker agent with standard command provided by server
>> no issue. connexion to websocket hoststats works correctly with a 101 reply

Host 3: fresh install of Ubuntu 15.10 (with KVM) (4.2.0-18-generic)
install docker agent with standard command provided by server
>> issue triggered. connexion to websocket through rancher UI triggers 401 error.
Accessing the same websocket directly (with chrome extension 'Simple WebSocket Client 0.1.3') correctly connect to socket

@vrparmar : which os/kernel are your hosts on ?

**Comment by dkiser** [comment]
+1 Same stack trace

**docker version**
```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64
```

**rancher version**
```
Component	Version
Rancher	v0.59.0
Cattle	v0.148.0
User Interface	v0.90.0
Rancher Compose	v0.7.2
```

**Comment by cloudnautique** [comment]
Can you show the Nginx ssl configuration? If it is trying to hit ws:// and not wss:// I suspect headers aren't being set/passed.

Also, it looks like our host api is failing pretty hard, but not able to talk to cAdvisor. On your nodes that are failing, do you see cAdvisor restarting as well?

**Comment by dkiser** [comment]
@cloudnautique - For my case at least, can't speak to OP

**nginx**
Configured correctly as 2/6 nodes in cluster have stats working and rancher-logs on hosts with stats broken even show `wss://` in the URI as well as developer tools in the browser.
```
time="2016-03-05T14:35:42Z" level="info" msg="Connecting to proxy." url="wss://SCRUBBED/v1/connectbackend?token=SCRUBBED_TOKEN"
```

**panic runtime**
Yes, cAdvisor is running and the logs show it, but the api is the one that keeps throwing the panic.  This is repeatable by refreshing the particular host's infrastructure page to attempt a reconnect for stats.  cAdvisor also stays bound to its port per below even after the GO panic.
```
[root@SRUBBED dkiser]# docker exec -it rancher-agent nc -vz localhost 9344
Connection to localhost 9344 port [tcp/*] succeeded!
```

**Comment by ibuildthecloud** [comment]
Woo hoo!  I like panics.  Those are easy to fix.  I'll look into this.

**Comment by dkiser** [comment]
@ibuildthecloud Thanks!  Let me know if I can gather any more info or want to screen share to hop in and look at state of things.

**Comment by wzrdtales** [comment]
@ibuildthecloud Experience the same problem.
On the same host as the rancher server currently running debian wheezy I do receive the stats, but on a different host, which are currently all debian jessie, didn't test yet with debian wheezy there, I get the exact same errors.

**Comment by wzrdtales** [comment]
@ibuildthecloud Have found something weird about this bug.

If you open a server from the hosts view, you get the websocket error. But if you choose a server from the dropdown **in** the hosts view and open another server from it, the stats are working again. And also if you switch back to the server via the drop down everything works.
Seems like only something in the call from the hosts view into the view of a host itself is not working properly.

**Comment by TheRubble** [comment]
I've noticed I only seem to get this with boxes on some hosts. I have 3 in Digital Ocean, 1 in Linode and a final cheap dev box running in kimsufi.

I cannot get any of the graphs working on the kimsufi box. I've checked firewalls etc.. but no luck.

Host : KimSufi
Kernel :  3.16.0-4-amd64
OS : Debian GNU/Linux 8 (jessie)

`/v1/hoststats?token=eyJhbGciOiJSUzI1NiJ9.eyJleHAiOjE0NTc3OTcwNTYsInN1YiI6ImNhdHRsZSIsInJlc291cmNlSWQiOiIxaDIyIiwiaXNzIjoiaHR0cDpcL1wvY2F0dGxlLmlvIiwiaG9zdFV1aWQiOiI3OTYzNzg0ZS00NDcyLTQ2ZGEtYjk0ZS05NGUzOWVjYjg4ZmYiLCJpYXQiOjE0NTc3OTY3NTYsImtpZCI6ImRlZmF1bHQifQ.EEVAZxp_wijOpnk2zHawv7SUBSeGqTjylgYRCV811Owpz3uzGnhRock5SAqnd4CDYGB2GSpckFyGJMwAfRTNw84buABzLJny31xYvaAa8FHlHTI2nuVUuOHCteyrTNWW8_7_90New_TXTUJzoHmcQiFpf-TSlqAe5SSugF4mSVuZL3cv-_OGrWRiajhmfzZ07E9pg1nB5ujke50Zuw6_uf_MHhD8bcVPmQbtUDN32vKllue1ZcA_TsVYi_GdYGL5zHNM6-vlG6hRNup3UGsReQWJaitghUiawT3t_2M28F7hRiAa9I6JaoI9eUmU829HFCH1cJ-2KPH5SaBa6pzCzA' failed: HTTP Authentication failed; no valid credentials available`

**Comment by wzrdtales** [comment]
Another bit of information from me:

Have tested it now. With debian wheezy, no problems. With debian jessie, this problem occurs. Seems like there are some incompatibilities in the new debian.

**Comment by clescot** [comment]
same issue with ubuntu server 15.10 and rancher:v0.63 on the kimsufi provider.
kernel :  3.14.32-xxxx-grs-ipv6-64 (kernel patched by kimsufi with security updates).
The trick with the drop down list works too.
Charles.

**Comment by dkiser** [comment]
@wzrdtales - confirmed the dropdown trick works for me on the CentOS setup reported earlier in the thread.

@ibuildthecloud - maybe a UUID for a resource got lost in translation between the infra views and the dropdown?

**Comment by brutus333** [comment]
I am experiencing exactly the same issue with Rancher v0.63.0, RedHat 7.2 and Docker 1.10.3. I confirm also the dropdown box workaround.
Any insight on when this will be fixed?

**Comment by brutus333** [comment]
I managed to spot a difference between hosts that have the graph shown without calling in the workaround and the ones that need the workaround while managed by the same Rancher instance (v 0.59.1): the ones that work have Docker 1.9x while the ones that don't have Docker 1.10.3.
The issue also appears on Rancher v0.63.0 with Docker 1.10.3 so it seems to be related somehow with the docker version.

**Comment by dkiser** [comment]
@brutus333 FYI I see the issue with all hosts on exact same docker 1.9.x version.

**Comment by ibuildthecloud** [comment]
We are now using a cadvisor v0.22 and additionally I've fixed the reported panic. https://github.com/rancher/host-api/pull/64

**Comment by sangeethah** [comment]
Please refer to https://github.com/rancher/rancher/issues/361 for list all the OS versions that were validated with cadvisor v0.22.
Fix for this issue is available in rancher-server version - v1.0.0-rc1.
